cmd/storagenode/main.go

```diff
@@ -88,6 +106,104 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
+	if err != nil {
+		return err
+	}
+
+	// check if the directory exists
+	_, err = os.Stat(diagCfg.BasePath)
+	if err != nil {
+		fmt.Println("Storagenode directory doesn't exist", diagCfg.BasePath)
+		return err
+	}
+
+	// open the sql db
+	dbpath := filepath.Join(diagCfg.BasePath, "piecestore.db")
+	db, err := psdb.Open(context.Background(), "", dbpath)
+	if err != nil {
+		fmt.Println("Storagenode database couldnt open:", dbpath)
+		return err
+	}
+
+	// Agreement is a struct that contains a bandwidth agreement and the associated signature
+	type SatAttributes struct {
+		TotalBytes        int64
+		PutActionCount    int64
+		GetActionCount    int64
+		TotalTransactions int64
+		// additional attributes add here ...
+	}
+
+	//get all bandwidth aggrements entries already ordered
+	bwAgreements, err := db.GetBandwidthAllocations()
+	if err != nil {
+		fmt.Println("bw allocation table read err")
+		return err
+	}
+
+	const padding = 3
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, padding, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	fmt.Fprintln(w, "SatelliteID\tTotal\t# Of Transactions\tPUT Action\tGET Action\t")
+
+	// attributes per satelliteid
+	satelliteID := make(map[string]SatAttributes)
+	satAtt := SatAttributes{}
+	var currSatID, lastSatID string
+
+	for _, v := range bwAgreements {
+		for _, j := range v {
+			// deserializing rbad you get payerbwallocation, total & storage node id
+			rbad := &pb.RenterBandwidthAllocation_Data{}
+			if err := proto.Unmarshal(j.Agreement, rbad); err != nil {
+				return err
+			}
+			total := rbad.GetTotal()
+
+			// deserializing pbad you get satelliteID, uplinkID, max size, exp, serial# & action
+			pbad := &pb.PayerBandwidthAllocation_Data{}
+			if err := proto.Unmarshal(rbad.GetPayerAllocation().GetData(), pbad); err != nil {
+				return err
+			}
+			currSatID = fmt.Sprintf("%s", pbad.GetSatelliteId())
+			action := pbad.GetAction()
+
+			if strings.Compare(currSatID, lastSatID) != 0 {
+				// make an entry
+				satAtt.TotalBytes = total
+				if action == pb.PayerBandwidthAllocation_PUT {
+					satAtt.PutActionCount++
+				} else {
+					satAtt.GetActionCount++
+					fmt.Println(satAtt.GetActionCount)
+				}
+				satAtt.TotalTransactions++
+				satelliteID[currSatID] = satAtt
+				lastSatID = currSatID
+			} else {
+				//update the already existing entry
+				for satIDKey, satIDVal := range satelliteID {
+					if strings.Compare(satIDKey, currSatID) == 0 {
+						satIDVal.TotalBytes = satIDVal.TotalBytes + total
+						if action == pb.PayerBandwidthAllocation_PUT {
+							satIDVal.PutActionCount++
+						} else {
+							satIDVal.GetActionCount++
+						}
+						satIDVal.TotalTransactions++
+						satelliteID[satIDKey] = satIDVal
+					}
+				}
+			}
+		}
+		fmt.Fprintln(w, currSatID, "\t", satelliteID[currSatID].TotalBytes, "\t", satelliteID[currSatID].TotalTransactions, "\t", satelliteID[currSatID].PutActionCount, "\t", satelliteID[currSatID].GetActionCount, "\t")
```

Comment:
println adds spaces between statements. Use Fprint with `\n` at the end.

Reply:
Also this breaks when the table is empty.

Reply:
modified as suggested